File: 2021/typescript/src/days/day4.ts
import {Day, sum} from "../aoc";

export class Day4 extends Day {
    day = (): number => 4;

    part1 = (): string => this.play(((currentBoard, calledNumber) => currentBoard.score(calledNumber).toString()))

    part2 = (): string =>
        this.play((currentBoard, calledNumber, boards) =>
            boards.length > 1
                ? boards.filter(x => x !== currentBoard)
                : currentBoard.score(calledNumber).toString())

    play(bingoProcedure: (currentBoard: Board, calledNumber: number, boards: Board[]) => string | Board[]): string {
        let boards = this.generateBoards();
        for(const n of this.allNumbers) {
            for(const currentBoard of boards) {
                currentBoard.call(n);
                if(currentBoard.hasBingo()) {
                    const res = bingoProcedure(currentBoard, n, boards);
                    if(typeof res === "string") return res;
                    boards = res;
                }
            }
        }
        return 'unknown';
    }

    private allNumbers: number[];
    private inputSegments: string[];

    setup() {
        super.setup();

        this.inputSegments = this.input.join("\n").split("\n\n");
        this.allNumbers = this.inputSegments.shift().split(",").map(Number);
    }

    private generateBoards = (): Board[] => this.inputSegments
        .map(b => b.split("\n").map(l => l.split(/\s+/).map(Number)))
        .map((b): Board => new Board(b));
}

class Board {
    all: number[];
    rows: number[][];

    constructor(numbers: number[][]) {
        this.all = numbers.flat();
        this.rows = [...numbers, ...Board.cols(numbers)];
    }

    public call(n: number): void {
        this.all = this.all.filter(x => x !== n);
        this.rows = this.rows.map(r => r.filter(x => x !== n));
    }

    public hasBingo = (): boolean => this.rows.find(r => r.length === 0) !== undefined;

    public score = (calledNumber: number): number => this.all.reduce(sum) * calledNumber;

    private static cols = (b: number[][]): number[][] => b[0].map((v, i) => b.map(x => x[i]));
}

File: 2021/julia/src/Day4.jl
module Day4

include("Aoc.jl")
using .Aoc

mutable struct Board
    square::Array{Int64}
    called::BitArray{2}
    done::Bool
    Board(square) = new(square, square .== -1)
end

input = Aoc.input_lines(4)
numbers = parse.(Int64, split(popfirst!(input), ','))
boards = input |>
    s -> strip(join(s, "\n")) |>
    s -> split(s, "\n\n") |>
    all -> map(board -> split.(split(board, "\n")), all) |>
    all -> map(board -> map(row -> parse.(Int64, row), board), all) |>
    all -> map(board -> permutedims(reduce(hcat, board)), all) |>
    all -> map(board -> Board(board), all)

call!(board::Board, number) = board.called = board.called .|| board.square .== number
score(board::Board, number) = sum(board.square .* (board.called .== 0)) * number
hasbingo(board::Board) = findfirst(r -> sum(r) == size(board.square, 1), [eachrow(board.called)..., eachcol(board.called)...]) != nothing

function part1()
    for n in numbers, b in boards
        call!(b, n)
        hasbingo(b) && return score(b, n)
    end
end

function part2()
    for n in numbers, b in boards
        b.done && continue
        call!(b, n)

        if hasbingo(b)
            count(b -> !b.done, boards) == 1 && return score(b, n)
            b.done = true
        end
    end
end

end